panimg/image_builders/oct.py

```diff
@@ -0,0 +1,214 @@
+from collections import defaultdict
+from pathlib import Path
+from typing import DefaultDict, Iterable, Iterator, List, Set, Tuple, Union
+
+import SimpleITK
+import numpy as np
+from construct.core import (
+    Float64l,
+    Int8ul,
+    PaddedString,
+    StreamError,
+    Struct,
+)
+from pydantic import BaseModel
+
+from panimg.contrib.oct_converter.image_types import (
+    FundusImageWithMetaData,
+    OCTVolumeWithMetaData,
+)
+from panimg.contrib.oct_converter.readers import E2E, FDA, FDS
+from panimg.exceptions import UnconsumedFilesException, ValidationError
+from panimg.models import EyeChoice, SimpleITKImage
+
+
+def format_error(message: str) -> str:
+    return f"OCT image builder: {message}"
+
+
+LATERALITY_TO_EYE_CHOICE = defaultdict(
+    lambda: EyeChoice.UNKNOWN,
+    {"L": EyeChoice.OCULUS_SINISTER, "R": EyeChoice.OCULUS_DEXTER},
+)
+
+OCT_CONVERTER_TYPE = Union[FDS, FDA, E2E]
+
+
+class OctSliceSpacing(BaseModel):
+    x_mm: float
+    y_mm: float
+    z_mm: float
+
+
+def create_itk_images(
+    *,
+    file: Path,
+    oct_volumes: Iterable[OCTVolumeWithMetaData],
+    fundus_images: Iterable[FundusImageWithMetaData],
+    oct_slice_size: OctSliceSpacing,
+) -> Iterator[SimpleITKImage]:
+    for volume in oct_volumes:
+        eye_choice = LATERALITY_TO_EYE_CHOICE[volume.laterality]
+        yield create_itk_oct_volume(
+            file=file,
+            volume=volume.volume,
+            oct_slice_size=oct_slice_size,
+            eye_choice=eye_choice,
+        )
+    for image in fundus_images:
+        eye_choice = LATERALITY_TO_EYE_CHOICE[image.laterality]
+
+        if file.suffix != ".e2e":
+            img_array = image.image.astype(np.uint8)
+            img_array = img_array[:, :, ::-1]
+            is_vector = True
```

Comment:
I'm not too sure if we can rely on checking the suffix here, maybe we need to try loading it as e2e and then adding the attributes we need for later. Also unsure about why we do the type conversion and invert the last axis.

Reply:
Ok! Now that we have included the oct converter in contrib we can make those fixes later, but not in this PR.